tortoise/sim/partition.go

```diff
@@ -0,0 +1,186 @@
+package sim
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// Frac is a shortcut for creating Fraction object.
+func Frac(nominator, denominator int) Fraction {
+	return Fraction{Nominator: nominator, Denominator: denominator}
+}
+
+// Fraction of something.
+type Fraction struct {
+	Nominator, Denominator int
+}
+
+func (f Fraction) String() string {
+	return fmt.Sprintf("%d/%d", f.Nominator, f.Denominator)
+}
+
+// SplitOpt is for configuring partition.
+type SplitOpt func(*splitConf)
+
+// WithPartitions configures number of miners in each partition
+// relative to original Generator.
+func WithPartitions(parts ...Fraction) SplitOpt {
+	return func(c *splitConf) {
+		c.Partitions = parts
+	}
+}
+
+func splitConfDefaults() splitConf {
+	return splitConf{
+		Partitions: []Fraction{Frac(1, 2), Frac(1, 2)},
+	}
+}
+
+type splitConf struct {
+	Partitions []Fraction
+	// TODO(dshulyak) support both transient and full partitions
+	//
+	// transient partitions can be configured without splitting state, in transient partitions
+	// - votes are splitted
+	// - hare will not reach consensus (or will reach only in one of the partition)
+	// - but atxs and beacons remain the same
+	//
+	// for now transient can be configured by options to Next
+}
+
+// Split generator into multiple partitions.
+// First generator will use original tortoise state (mesh, activations).
+//
+// This has an issue, original beacon is preserved only in the first generator
+// therefore rerun after merge will work only in the first generator.
+func (g *Generator) Split(opts ...SplitOpt) []*Generator {
+	// Split everything that would be changed after long network partition
+	// - activations
+	// - beacons
+	// - number of miners in each partition
+	//
+	//  things that should remain the same:
+	// - blocks, contextually valid blocks and input vectors before partition
+	// - activations before partition
```

Comment:
aren't beacons before partition also unchanged?

Reply:
yes they should, i will mention them